src/Compilers/Core/Portable/Symbols/Attributes/CustomAttributesBag.cs

```diff
@@ -68,6 +68,8 @@ public bool IsEmpty
         public bool SetEarlyDecodedWellKnownAttributeData(EarlyWellKnownAttributeData data)
         {
             WellKnownAttributeData.Seal(data);
+            // Early decode must complete before full decode
+            Debug.Assert(!IsPartComplete(CustomAttributeBagCompletionPart.DecodedWellKnownAttributeData));
             var setOnOurThread = Interlocked.CompareExchange(ref _earlyDecodedWellKnownAttributeData, data, null) == null;
             NotePartComplete(CustomAttributeBagCompletionPart.EarlyDecodedWellKnownAttributeData);
```

Comment:
>Debug.Assert(!IsPartComplete(CustomAttributeBagCompletionPart.DecodedWellKnownAttributeData)); [](start = 12, length = 94)

Perhaps instead of removing the assert it could be changed to:
```
Debug.Assert(!IsPartComplete(CustomAttributeBagCompletionPart.DecodedWellKnownAttributeData) || IsPartComplete(CustomAttributeBagCompletionPart.EarlyDecodedWellKnownAttributeData));
```

Reply:
Good suggestion. I think I've convinced myself that this is safe 😄 

I was concerned about potential read reordering issues but these are just loading the same block of memory and the state changes are atomic on that same piece of memory. Hence this should be robust to that type of problem.